### PR TITLE
Queue

### PR DESCRIPTION
**Pull request by cameronhudson1**
Added basic functionality for managing queue in the backend.

Songs can be added and voted on through calls to the queue object.  The Song instances keep track of their current actually i dont care u know what it does

### CODE REVIEW

File: src/queue/song.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class Song:
+
+    def __init__(self, code, votes, ip):
+        self.code = code
+        self.votes = votes
+        self.upvoted = [ip]
```
Comment (carsonclarke570) [inline src/queue/song.py:6]: Is this assuming when someone adds a song that they are also upvoting it?

File: src/queue/song.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+class Song:
+
+    def __init__(self, code, votes, ip):
+        self.code = code
+        self.votes = votes
```
Comment (carsonclarke570) [inline src/queue/song.py:5]: This might be redundant. We can get the votes by counting the number upvotes and downvotes

File: src/queue/sq.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    def add(self, code, ip):
+        """
+        Adds a new song to the queue
+        :param code: spotify code of song to add
+        :param ip: ip of user who added this song
+        :return:
+        """
+        new_song = Song(code, 1, ip)
+        assert (new_song not in self.queue), "No duplicates allowed in queue"
```
Comment (carsonclarke570) [inline src/queue/sq.py]: started thinking about assert statements and realized it's better if we just throw an error. The handler classes has built in error catching so we can freely throw exceptions wherever the fuck we want. Look at the HandlerException class for an example exception

File: src/queue/sq.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        self.queue.append(new_song)
+        self.update()
+
+    def remove(self, code):
```
Comment (carsonclarke570) [inline src/queue/sq.py:39]: Donde esta

File: src/queue/sq.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    def get_queue(self):
+        """
+        Returns the queue
+        :return: the song queue data structure
+        """
+        return self.queue
+
+
+def test():
```
Comment (carsonclarke570) [inline src/queue/sq.py:98]: we can leave this in for now but we'll convert it to unit tests later

**Review by carsonclarke570 — CHANGES_REQUESTED** [review]
Looks mostly good. A few small fixes for robustness, bitch

File: src/queue/song.py (hunk 1 above)
Comment (cameronhudson1) [reply src/queue/song.py:6]: Yes, that's the implementation I suggest for the time being.  Similar to reddit for upvoting.

File: src/queue/song.py (hunk 2 above)
Comment (cameronhudson1) [reply src/queue/song.py:5]: Faster access speeds for low iterative cost.  I think it's beneficial to include.

File: src/queue/sq.py (hunk 3 above)
Comment (cameronhudson1) [reply src/queue/sq.py]: Sure, I can convert to errors.  Any format for how you want it?

File: src/queue/sq.py (hunk 5 above)
Comment (cameronhudson1) [reply src/queue/sq.py:98]: Yep, sounds good.

**Review by carsonclarke570 — APPROVED** [review]

**Pull request merged**